Fix kernel cmdline no null terminal issue.

Meanwhile explicitly enlarge the cmdline size to 1023 bytes

Signed-off-by: Gang Wei <gang.wei@intel.com>

--- tboot/common/linux.c.orig	2014-05-16 03:57:00.000000000 -0400
+++ tboot/common/linux.c	2014-05-28 15:07:58.278772343 -0400
@@ -121,7 +121,7 @@
     /* recommended layout
         0x0000 - 0x7FFF     Real mode kernel
         0x8000 - 0x8FFF     Stack and heap
-        0x9000 - 0x90FF     Kernel command line
+        0x9000 - 0x93FF     Kernel command line
         for details, see linux_defns.h
     */
 
@@ -144,13 +144,13 @@
     if ( hdr->header != HDRS_MAGIC ) {
         /* old kernel */
         printk(TBOOT_ERR
-               "Error: Old kernel (< 2.6.20) is not supported by tboot.\n");
+               "Error: Old kernel (< 2.6.22) is not supported by tboot.\n");
         return false;
     }
 
-    if ( hdr->version < 0x0205 ) {
+    if ( hdr->version < 0x0206 ) {
         printk(TBOOT_ERR
-               "Error: Old kernel (<2.6.20) is not supported by tboot.\n");
+               "Error: Old kernel (<2.6.22) is not supported by tboot.\n");
         return false;
     }
 
@@ -287,6 +287,7 @@
 
     /* set cmd_line_ptr */
     hdr->cmd_line_ptr = real_mode_base + KERNEL_CMDLINE_OFFSET;
+    hdr->cmdline_size = REAL_END_OFFSET - KERNEL_CMDLINE_OFFSET - 1;
 
     /* load protected-mode part */
     memmove((void *)protected_mode_base, linux_image + real_mode_size,
@@ -307,7 +308,8 @@
     printk(TBOOT_INFO"Linux cmdline placed in header: ");
     printk_long(kernel_cmdline);
     printk(TBOOT_INFO"\n");
-    memcpy((void *)hdr->cmd_line_ptr, kernel_cmdline, strlen(kernel_cmdline));
+    strncpy((void *)hdr->cmd_line_ptr, kernel_cmdline, hdr->cmdline_size);
+    ((char *)hdr->cmd_line_ptr)[hdr->cmdline_size] = '\0';
 
     /* need to put boot_params in real mode area so it gets mapped */
     boot_params = (boot_params_t *)(real_mode_base + real_mode_size);
--- tboot/include/linux_defns.h.orig	2014-05-16 03:57:00.000000000 -0400
+++ tboot/include/linux_defns.h	2014-05-28 15:01:36.643788514 -0400
@@ -178,7 +178,7 @@
         |  I/O memory hole          |
 0A0000  +---------------------------+
         |  Reserved for BIOS        |  Do not use.  Reserved for BIOS EBDA.
-099100  +---------------------------+
+099400  +---------------------------+
         |  cmdline                  |
 099000  +---------------------------+
         |  Stack/heap               |  For use by the kernel real-mode code.
@@ -203,7 +203,7 @@
 #define REAL_KERNEL_OFFSET      0x0000
 #define BOOT_SECTOR_OFFSET      0x0200
 #define KERNEL_CMDLINE_OFFSET   0x9000
-#define REAL_END_OFFSET         0x9100
+#define REAL_END_OFFSET         0x9400
 
 #define REAL_MODE_SIZE          REAL_END_OFFSET - REAL_KERNEL_OFFSET
 
